### PR TITLE
Add workaround for GWCS error in some JWST IFU cubes

### LINKED ISSUE #956: JWST s3d loader broken for latest cube data
The specutils loader for JWST s3d files is broken for recent NIRSpec IFU cube data cal_ver 1.4.6.   It gives the error `ValueError: Spectral axis length (973) must be the same size or one greater (if specifying bin edges) than that of the last flux axis (950)`.  I think there's a mismatch between the input data and the loader code to extract WCS and convert it to a wavelength. 

To reproduce:
```
data = "jw01409-o018_t011_nirspec_prism-clear_s3d.fits"
ss = Spectrum1D.read(data)
```
 
Traceback:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [24], in <cell line: 1>()
----> 1 ss=Spectrum1D.read(hdu.filename())

File ~/anaconda3/envs/vmpy3.9/lib/python3.9/site-packages/astropy/nddata/mixins/ndio.py:59, in NDDataRead.__call__(self, *args, **kwargs)
     58 def __call__(self, *args, **kwargs):
---> 59     return self.registry.read(self._cls, *args, **kwargs)

File ~/anaconda3/envs/vmpy3.9/lib/python3.9/site-packages/astropy/io/registry/core.py:199, in UnifiedInputRegistry.read(self, cls, format, cache, *args, **kwargs)
    195     format = self._get_valid_format(
    196         'read', cls, path, fileobj, args, kwargs)
    198 reader = self.get_reader(format, cls)
--> 199 data = reader(*args, **kwargs)
    201 if not isinstance(data, cls):
    202     # User has read with a subclass where only the parent class is
    203     # registered.  This returns the parent class, so try coercing
    204     # to desired subclass.
    205     try:

File ~/anaconda3/envs/vmpy3.9/lib/python3.9/site-packages/specutils/io/default_loaders/jwst_reader.py:531, in jwst_s3d_single_loader(filename, **kwargs)
    513 @data_loader(
    514     "JWST s3d", identifier=identify_jwst_s3d_fits, dtype=Spectrum1D,
    515     extensions=['fits'], priority=10,
    516 )
    517 def jwst_s3d_single_loader(filename, **kwargs):
    518     """
    519     Loader for JWST s3d 3D rectified spectral data in FITS format.
    520
   (...)
    529         The spectrum contained in the file.
    530     """
--> 531     spectrum_list = _jwst_s3d_loader(filename, **kwargs)
    532     if len(spectrum_list) == 1:
    533         return spectrum_list[0]

File ~/anaconda3/envs/vmpy3.9/lib/python3.9/site-packages/specutils/io/default_loaders/jwst_reader.py:618, in _jwst_s3d_loader(filename, **kwargs)
    615         mask_name = primary_header.get("MASKEXT", "DQ")
    616         mask = hdulist[mask_name].data.T
--> 618         spec = Spectrum1D(flux=flux, spectral_axis=wavelength, meta=meta,
    619                           uncertainty=err, mask=mask)
    620         spectra.append(spec)
    622 return SpectrumList(spectra)

File ~/anaconda3/envs/vmpy3.9/lib/python3.9/site-packages/specutils/spectra/spectrum1d.py:178, in Spectrum1D.__init__(self, flux, spectral_axis, wcs, velocity_convention, rest_value, redshift, radial_velocity, bin_specification, **kwargs)
    176         bin_specification = "edges"
    177     else:
--> 178         raise ValueError(
    179             "Spectral axis length ({}) must be the same size or one "
    180             "greater (if specifying bin edges) than that of the last "
    181             "flux axis ({})".format(spectral_axis.shape[0],
    182                                     flux.shape[-1]))
    184 # If a WCS is provided, check that the spectral axis is last and reorder
    185 # the arrays if not
    186 if wcs is not None and hasattr(wcs, "naxis"):

ValueError: Spectral axis length (973) must be the same size or one greater (if specifying bin edges) than that of the last flux axis (950)
```

### PR DESCRIPTION
**Pull request by rosteen**
Closes #956 with a workaround until the pipeline team fixes the underlying GWCS problem. MJD-OBS code taken from [jdaviz](https://github.com/spacetelescope/jdaviz/blob/70eb5f16aa78e8de68faa3c7b8afbb7096f067ad/jdaviz/configs/cubeviz/plugins/parsers.py#L125), which solved a similar issue there.

### CODE REVIEW

**Comment by rosteen** [comment]
@pllim I tagged you for a review since you have access to the file that was failing before. If you need me to resend it/link you to it let me know.

File: specutils/io/default_loaders/jwst_reader.py [hunk 1]
```diff
@@ -585,6 +587,24 @@ def _jwst_s3d_loader(filename, **kwargs):
 
             wavelength = Quantity(wavelength_array, unit=wavelength_unit)
 
+            # The GWCS is currently broken for some IFUs, here we work around that
+            wcs = None
+            if wavelength.shape[0] != flux.shape[-1]:
+                # Need MJD-OBS for this workaround
+                if 'MJD-OBS' not in hdu.header:
```
Comment (pllim) [inline specutils/io/default_loaders/jwst_reader.py:594]: Might no longer be necessary with https://github.com/astropy/astropy/pull/12598 in astropy now, though I guess does not hurt to get this working with older astropy versions.

File: specutils/io/default_loaders/jwst_reader.py [hunk 2]
```diff
@@ -585,6 +587,24 @@ def _jwst_s3d_loader(filename, **kwargs):
 
             wavelength = Quantity(wavelength_array, unit=wavelength_unit)
 
+            # The GWCS is currently broken for some IFUs, here we work around that
```
Comment (pllim) [inline specutils/io/default_loaders/jwst_reader.py:590]: Are you sure this affects GWCS and not the FITS WCS?

**Review by pllim — APPROVED** [review]
The file loads now but I cannot tell if it is from this patch or from dev astropy that I have installed. I think it is best of @havok2063 tries to load the file too with whatever astropy that MAST is using.

**Comment by pllim** [comment]
p.s. Also works with astropy 5.1 but MAST should still test this, just in case.

File: specutils/io/default_loaders/jwst_reader.py (hunk 2 above)
Comment (rosteen) [reply specutils/io/default_loaders/jwst_reader.py:590]: Yes, the root issue is that the GWCS is specifying an array size that doesn't match the actual data array size along the spectral axis.

**Comment by havok2063** [comment]
I just tested this locally with 
```
data = "jw01409-o018_t011_nirspec_prism-clear_s3d.fits"
ss = Spectrum1D.read(data)
```
and it now works.  Locally I'm using `astropy 5.0.4`.  Our MAST environment is using the same astropy version.

**Comment by rosteen** [comment]
@havok2063 thanks for checking!

**Pull request merged**